compiler/sem/semstmts.nim

```diff
@@ -2633,15 +2639,30 @@ proc semProcAux(c: PContext, n: PNode, kind: TSymKind,
   # way of pragmas, default params, and so on invalidate this parsing.
   # Nonetheless, we need to carry out this analysis to perform the search for a
   # potential forward declaration.
-  setGenericParamsMisc(c, n)
 
-  s.typ =
-    if n[paramsPos].kind != nkEmpty:
-      semParamList(c, n[paramsPos], n[genericParamsPos], s.kind)
+  case generic.kind
+  of nkEmpty:
+    routine[genericParamsPos] = newNodeI(nkGenericParams, routine.info)
+  of nkGenericParams:
+    # we keep the original params around for better error messages, see
+    # issue https://github.com/nim-lang/Nim/issues/1713
+    routine[genericParamsPos] = semGenericParamList(c, generic)
+  else:
+    unreachable(generic.kind)
+
+  # TODO: refactor ``semParamList`` to not modify the input parameter list
+  routine[paramsPos] = formal
+
+  result =
+    case formal.kind
+    of nkFormalParams:
+      semParamList(c, routine[paramsPos], routine[genericParamsPos], kind)
+    of nkEmpty:
+      nil # let the callee handle empty parameter lists
```

Comment:
since we have the `routine` we can produce the correct info, why not avoid the `nil` return by making the type here?

Reply:
It should have been `caller` instead of `callee`, sorry.

As for avoiding `nil`, I don't see how that would work outside of resorting to conditional logic in `semRoutineParams`. Templates treat a missing parameter list differently than a macro or other routine.

Reply:
My bad I missed the difference in the template callsite.